File: src/main/java/com/my/admin/app/commerce/service/UserService.java
package com.my.admin.app.commerce.service;

import com.black.food.manager.dao.UserDao;
import com.black.food.manager.model.api.response.UserResponse;
import com.black.food.manager.model.entity.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    private UserDao userDao;

    public List<User> findAll() { return userDao.findAll(); }

    public UserResponse getUserResponseById(final Long id){
        return userDao.getUserResponseById(id).orElseThrow();
    }

    public Optional<UserResponse> findByUserName(final String userName) { return userDao.findByUserName(userName); }

    @Autowired
    public void setUserDao(final UserDao userDao) {
        this.userDao = userDao;
    }
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/service/UserService.kt
package com.bk.reactive.app.commerce.admin.my.service

import com.bk.reactive.app.commerce.admin.my.dao.UserDAO
import com.bk.reactive.app.commerce.admin.my.exception.api.ApplicationException
import com.bk.reactive.app.commerce.admin.my.model.api.response.UserResponse
import com.bk.reactive.app.commerce.admin.my.model.mapper.UserMapper.Companion.fromUserToUserResponse
import mu.KotlinLogging
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class UserService(
    private val userDAO: UserDAO
) {
    private val logger = KotlinLogging.logger { }

    fun findAll(): Flux<UserResponse> = userDAO.findAll()
        .map { fromUserToUserResponse(it) }

    fun findByUserName(userName: String): Mono<UserResponse> {
        return userDAO.findByUserName(userName)
            .map { fromUserToUserResponse(it) }
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error {
                    ApplicationException(message = it.localizedMessage, cause = Throwable(it.localizedMessage))
                }
            }

    }

    fun getUserResponseById(id: String): Mono<UserResponse> {
        return userDAO.getUserById(id)
            .map { fromUserToUserResponse(it) }
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error {
                    ApplicationException(message = it.localizedMessage, cause = Throwable(it.localizedMessage))
                }
            }
    }
}